src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -102,86 +80,68 @@ BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hint
     }
 
     // Push 1 blk to the vector which has all the requested nblks;
-    out_bid = BlkId{m_last_append_offset, nblks, m_chunk_id};
-
-    auto cur_cp = hs()->cp_mgr().cp_guard();
-    m_last_append_offset += nblks;
-    m_freeable_nblks -= nblks;
-
-    // it is guaranteed that dirty buffer always contains updates of current_cp or next_cp, it will
-    // never get dirty buffer from across updates;
-    set_dirty_offset(cur_cp->id() % MAX_CP_COUNT);
+    out_bid = BlkId{m_last_append_offset.fetch_add(nblks), nblks, m_chunk_id};
 
     COUNTER_INCREMENT(m_metrics, num_alloc, 1);
 
     return BlkAllocStatus::SUCCESS;
 }
 
-BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-BlkAllocStatus AppendBlkAllocator::mark_blk_allocated(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-void AppendBlkAllocator::free_on_disk(BlkId const&) {}
+BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) {
+    m_is_dirty.store(true);
+    return BlkAllocStatus::SUCCESS;
+}
 
-bool AppendBlkAllocator::is_blk_alloced_on_disk(BlkId const&, bool) const { return false; }
+BlkAllocStatus AppendBlkAllocator::alloc_on_cache(BlkId const& blkid) {
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_last_append_offset.load();
+    do {
+        if (cur_offset >= new_offset) { break; } // Already allocated
+    } while (!m_last_append_offset.compare_exchange_weak(cur_offset, new_offset));
+    return BlkAllocStatus::SUCCESS;
```

Comment:
1 m_is_dirty.store(true)？ like my above comment

2 if a write request comes when replaying log (which is carried out in a background thread and will commit the blk), this chunk might be selected for this write . then  a new blk will be allocated from this chunk(`AppendBlkAllocator::alloc` is called.) 
if `AppendBlkAllocator::alloc` is called between line 97 and 98 , `m_last_append_offset` will be a new one, and thus `m_last_append_offset.compare_exchange_weak(cur_offset, new_offset)` will fail for ever, right?

Reply:
1. As responded above.
2. `m_last_append_offset.compare_exchange_weak(cur_offset, new_offset)` will not forever fail, because compare_exchange_weak will set the cur_offset with whatever the m_last_append_offset value. This is done by the std::atomic::compare_exchange_weak to avoid additional atomic operation, reference: https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

Atomically compares the [object representation](https://en.cppreference.com/w/cpp/language/object)(until C++20)[value representation](https://en.cppreference.com/w/cpp/language/object)(since C++20) of *this with that of expected. If those are bitwise-equal, replaces the former with desired (performs read-modify-write operation). Otherwise, loads the actual value stored in *this into expected (performs load operation).

Reply:
thanks for the explanation of `compare_exchange_weak`, I do misunderstand it, my bad.
but this risk I mentioned above is still possible. I believe  we do not have any mechanism to keep `append_blk_allocator` from handling `allocate_blk` reqeust before all the logs are replayed. right?

for example , if we get homestore restarted and replaying log , when reaching line 97, we get an m_last_append_offset of 100. if an alloc request comes at this moment, maybe m_last_append_offset will be set to 101. but 101 might be already allocated to other blk before restarting. that`s the case i am saying.

so does homestore start handling write request only until all the logs are replyed? this is my point

Reply:
Yes homestore start will not be available to accept writes or follower commits until we finish replay.

Reply:
how to achieve this goal?
if that , reft_repl_service#start() should return only until all the repl_dev has replayed all the logs to the latest commit index. this way, it can block sm from accepting new writes.

can you point out where this part of code is?

Reply:
In RaftReplService::start, we load all repl_devs here
https://github.com/eBay/HomeStore/blob/master/src/lib/replication/service/raft_repl_service.cpp#L106

Then we start logstore_service.start() which replays all logs (which at this point repl_dev would replay till checkpoint lsn)

Then only we join the raft_group. After raft_group has been joined raft will commit things in lsn order only. So there is never a chance of out-of-order commit or action.

Reply:
sorry , i do not get your point here. logstore_service.start() will only open devs , right?  log replay should involve the statemachine of repl_dev, which will apply the change in log entry.

so can you point out the code where is this part ? i am a little confused here